File: 2021_Dev_Matching_First_Half/GoLang/77485.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func testing_77485() {
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()
	fmt.Fprintln(writer, solution_77485(6, 6, [][]int{{2, 2, 5, 4}, {3, 3, 6, 6}, {5, 1, 6, 3}}))
	fmt.Fprintln(writer, solution_77485(3, 3, [][]int{{1, 1, 2, 2}, {1, 2, 2, 3}, {2, 1, 3, 2}, {2, 2, 3, 3}}))
	fmt.Fprintln(writer, solution_77485(100, 97, [][]int{{1, 1, 100, 97}}))
}

func solution_77485(rows int, columns int, queries [][]int) []int {
	result := make([]int, len(queries))
	board := make([][]int, rows)

	for i := 0; i < rows; i++ {
		boardTmp := make([]int, columns)
		for j := 0; j < columns; j++ {
			boardTmp[j] = (i * columns) + (j + 1)
		}
		board[i] = boardTmp
	}

	for num, query := range queries {
		for i := range query {
			query[i] -= 1
		}

		tmpData := board[query[0]][query[1]]
		min := tmpData

		for i := query[1] + 1; i <= query[3]; i++ {
			tmpData, min = boardMove(board, query[0], i, tmpData, min)
		}
		for i := query[0] + 1; i <= query[2]; i++ {
			tmpData, min = boardMove(board, i, query[3], tmpData, min)
		}
		for i := query[3] - 1; i >= query[1]; i-- {
			tmpData, min = boardMove(board, query[2], i, tmpData, min)
		}
		for i := query[2] - 1; i >= query[0]; i-- {
			tmpData, min = boardMove(board, i, query[1], tmpData, min)
		}

		// 정상적인 알고리즘 (두번째 생각한거..)
		/*
			for i := query[0]; i < query[2]; i++ {
				val := board[i+1][query[1]]
				board[i][query[1]] = board[i+1][query[1]]
				if val < min {
					min = val
				}
			}
			for i := query[1]; i < query[3]; i++ {
				val := board[query[2]][i+1]
				board[query[2]][i] = board[query[2]][i+1]
				if val < min {
					min = val
				}
			}
			for i := query[2]; i > query[0]; i-- {
				val := board[i-1][query[3]]
				board[i][query[3]] = board[i-1][query[3]]
				if val < min {
					min = val
				}
			}
			for i := query[3]; i > query[1]; i-- {
				val := board[query[0]][i-1]
				board[query[0]][i] = board[query[0]][i-1]
				if val < min {
					min = val
				}
			}
			board[query[0]][query[1]+1] = tmpData
		*/

		result[num] = min
	}

	return result
}

func boardMove(board [][]int, q1 int, q2 int, tmpData int, min int) (int, int) {
	dummyData := board[q1][q2]
	board[q1][q2] = tmpData

	if dummyData < min {
		min = dummyData
	}

	return dummyData, min
}

File: 2021_Dev_Matching_First_Half/CPP/77485.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    vector<int> answer(queries.size(), 0);
    vector<vector<int>> board(rows, vector<int>(columns, 0));

	for (int i = 0; i < rows; i++) {
		for (int j = 0; j < columns; j++) {
			board[i][j] = (i * columns) + (j + 1);
		}
	}

    int now = 0;
    for (vector<int> query : queries) {
        for (int i = 0; i < query.size(); i++) {
            query[i] -= 1;
        }

        int tmpData = board[query[0]][query[1]];
		int min = tmpData;

        for (int i = query[0]; i < query[2]; i++) {
			int val = board[i+1][query[1]];
			board[i][query[1]] = val;
			if (val < min) {
				min = val;
			}
		}

		for (int i = query[1]; i < query[3]; i++) {
			int val = board[query[2]][i+1];
			board[query[2]][i] = val;
			if (val < min) {
				min = val;
			}
		}

		for (int i = query[2]; i > query[0]; i--) {
			int val = board[i-1][query[3]];
			board[i][query[3]] = val;
			if (val < min) {
				min = val;
			}
		}

		for (int i = query[3]; i > query[1]; i--) {
			int val = board[query[0]][i-1];
			board[query[0]][i] = val;
			if (val < min) {
				min = val;
			}
		}
        board[query[0]][query[1]+1] = tmpData;
        
        answer[now] = min;
        now++;
    }

    return answer;
}

int main() {
    vector<vector<int>> q1({ {2, 2, 5, 4}, {3, 3, 6, 6}, {5, 1, 6, 3} });
    for (int i : solution(6, 6, q1)) {
        cout << i << " ";
    }

    cout << endl;

    vector<vector<int>> q2({ {1, 1, 2, 2}, {1, 2, 2, 3}, {2, 1, 3, 2}, {2, 2, 3, 3} });
    for (int i : solution(3, 3, q2)) {
        cout << i << " ";
    }
    
    return 0;
}